### PR TITLE
setup restaurant-header-hours-atom

### PR DESCRIPTION
**Pull request by mjedwards**

### CODE REVIEW

File: src/ui/views/restaurantView/restaurantView.jsx [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+// if CT is greater than open and less than close than opened
+// if CT is greater than open and greater than close than closed
+
+// verify the current time by using the length. we know if current time is 4 than we use "is close time less than greater than open" else if the length of current time is 3 than we stick to the orginal operation
+
+function RestaurantView() {
+    let timeSlots = [];
+    let restarauntHours = '';
+    const allDays = ["sun", "mon", "tue", "wed", "thu", "fri", "sat"];
+    const today = date.toUTCString();
```
Comment (MatthewVaccaro) [inline src/ui/views/restaurantView/restaurantView.jsx]: `const today = date.toUTCString();`

^ We won't be able to use this because it's too prescriptive. The Method being used "UTCString()" takes a date object and changes its time zone to be equal with the UTC timezone. When checking time we need to use the user's timezone.

File: src/ui/views/restaurantView/restaurantView.jsx [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    const day = date.getDay();
```
Comment (MatthewVaccaro) [inline src/ui/views/restaurantView/restaurantView.jsx]: `const date = new Date();` === `const day = date.getDay();` 

Line 16 and 40 do the same thing

File: src/ui/views/restaurantView/restaurantView.jsx [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+const restaurantOne = {
+    "id": 17,
+    "restaurant_ref": 17,
+    "mon": "1:00 PM 1:30 AM",
+    "tue": "10:00 AM 8:30 PM",
+    "wed": "10:00 PM 4:30 AM",
+    "thu": "10:00 AM 8:30 PM",
+    "fri": "2:00 PM 10:30 PM",
+    "sat": "10:00 AM 8:30 PM",
```
Comment (MatthewVaccaro) [inline src/ui/views/restaurantView/restaurantView.jsx]: Code broke when testing with Saturday (the day I am reviewing this) is set to "Closed"

File: src/ui/views/restaurantView/restaurantView.jsx [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        let addTwelve = parseInt(arr[0])+12
+        arr[0] = addTwelve.toString()
+        let fullTime = arr[0]+arr[1]
+        let disfullTIme = parseInt(fullTime)
+        return disfullTIme
+    }
+    
+
+    let amOrpm = () =>{
+        let time = restaurantOne[localTimeArr[0].toLowerCase()].split(' ')
```
Comment (MatthewVaccaro) [inline src/ui/views/restaurantView/restaurantView.jsx]: I understand what this code is doing but I don't understand the utility. It makes sense to be checking if the ending time is AM or PM but what I don't understand is why it's all being added to an array called time slots. You wouldn't need AM or PM if it's set in Military time, correct? It feels like a lot of extra logic to convert 12hour time into 24hour time

Heres an alternative to finding the 24Hour Converstions
`const openHour = time[1] === "AM ? (stringToInt(time[0]) : (stringToInt(time[0]) + 1200;`
You would do this for both the open and closed time.

File: src/ui/views/restaurantView/restaurantView.jsx [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    // removes the "":"":seconds from the array
+    currentTimeArray.pop()
+    // concats two strings in array to 24 hour string of just hour and minutes
+    let twentyFourHourTime = currentTimeArray[0] + currentTimeArray[1]
+    // sets the first index equal to the concatenated string
+    currentTimeArray[0] = twentyFourHourTime
+    // removes the last index of array leaving the 24 hour time
+    currentTimeArray.pop()
+    // turns 24 hour string to integer
+    let completeTime = parseInt(currentTimeArray[0])
```
Comment (MatthewVaccaro) [inline src/ui/views/restaurantView/restaurantView.jsx]: I understand what's happening between `105 - 115` but again feels like a lot of logic.

Line 107: I understand the seconds aren't going to be used but this line is extra code that's a bit unnecessary
Line 111: Why re-update this value? Does this string need to be held in an array? Are you planning on looping over it? 
Line 113: I understand wanting to clean up the array but again do you need this string to be in an array? 
Line 115: This nearly invalidates all of the lines above

This is a lot of extra code to get one simple thing done. These extra steps may not have an effect on time complexity but it does make it harder to understand your code and your purpose.

Recommendation: 
`Const completeTime = parseInt( currentTimeArray[0] + currentTimeArray[1] )`
This will achieve all of the lines above with far less code and steps to follow along

File: src/ui/views/restaurantView/restaurantView.jsx [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+
+    let timeRange;
+
+    let setHours = () => {
+        let currentTimeStr = completeTime.toString()
+        let timeSlotStrOpen = timeSlots[0].toString()
+        let timeSlotStrClose = timeSlots[1].toString()
```
Comment (MatthewVaccaro) [inline src/ui/views/restaurantView/restaurantView.jsx]: This line is what breaks your code when the object above "restaurant" has a day that is "closed". Line 122 is looking for the second item, but on a day when it's closed there is only one item.

File: src/ui/views/restaurantView/restaurantView.jsx [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+        return disfullTIme
+    }
+    
+
+    let amOrpm = () =>{
+        let time = restaurantOne[localTimeArr[0].toLowerCase()].split(' ')
+        console.log(time)
+        if (restaurantOne[localTimeArr[0].toLowerCase()] === "Closed") {
+            timeSlots.push("closed")
+            console.log(timeSlots)
```
Comment (MatthewVaccaro) [inline src/ui/views/restaurantView/restaurantView.jsx]: When you know the restaurant is closed you need to end the logic there. With how things are currently written, the logic below will still run.

Another problem you'll run into is you don't check the previous day.

Example:
The restaurant is closed all day Saturday. At exactly midnight the app would then say, it's Saturday, thus the restaurant is closed. What if on Fridays the restaurant doesn't close until 3 am? 

Restaurant hours transcends days. Your logic must check the past restaurant day to see if it's still open or not.

File: src/ui/views/restaurantView/restaurantView.jsx [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+    let setHours = () => {
+        let currentTimeStr = completeTime.toString()
+        let timeSlotStrOpen = timeSlots[0].toString()
+        let timeSlotStrClose = timeSlots[1].toString()
+        console.log(completeTime, "set hours function", currentTimeStr, timeSlotStrClose)
+        // checking the state of the time 
+        if (timeSlotStrClose.length === 3 && currentTimeStr.length === 4) {
+            if (completeTime < timeSlots[0]) {
+                console.log('close', 1, timeSlots[0])
+                restarauntHours = 'close'
```
Comment (MatthewVaccaro) [inline src/ui/views/restaurantView/restaurantView.jsx]: I left a comment about this above but, you need to be checking the previous day to ensure its not over

File: src/ui/views/restaurantView/restaurantView.jsx [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+            } 
+            if (completeTime > timeSlots[0]) {
+                console.log('open', 3, timeSlots[0])
+                restarauntHours = 'open'
+            }  
+        }
+        if (timeSlotStrClose.length === 3 && currentTimeStr.length === 3) {
+            if (completeTime < timeSlots[1]) {
```
Comment (MatthewVaccaro) [inline src/ui/views/restaurantView/restaurantView.jsx]: This doesn't factor in if the time the restaurant is open has passed or not. Its only looking to see if the current time has passed the closed time for a restaurant

File: src/ui/views/restaurantView/restaurantView.jsx [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+                console.log('open', 3, timeSlots[1])
+                restarauntHours = 'open'
+            }  
+            if (completeTime > timeSlots[1]) {
+                console.log('close', 4, timeSlots[1])
+                restarauntHours = 'close'
+            }  
+        } 
+
+        if (timeSlotStrOpen.length === 3 && currentTimeStr.length === 3){
```
Comment (MatthewVaccaro) [inline src/ui/views/restaurantView/restaurantView.jsx]: Same logic as above with a different outcome -- I am honestly very confused

File: src/ui/views/restaurantView/restaurantView.jsx [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+            if (completeTime < timeSlots[0]) {
+                    console.log('close')
+                    restarauntHours = 'close'
+                }
+                if (completeTime > timeSlots[0]) {
+                    console.log('open')
+                    restarauntHours = 'open'
+                }
+        }
+        if (timeSlotStrOpen.length === 3 && currentTimeStr.length === 4){
```
Comment (MatthewVaccaro) [inline src/ui/views/restaurantView/restaurantView.jsx]: Isn't this the same as the first `IF` clause above?

File: src/ui/views/restaurantView/restaurantView.jsx [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+            <div className='font-semibold bg-darkGray text-center rounded-md'>   
+                {restarauntHours === 'Closed' ? `Closed` : restarauntHours === 'open' ? <div><span className='text-green'>Open</span>{` Until ${(parseInt(timeRange[2]))+' '+ timeRange[3]}`} </div>: <div><span className='text-red'>Closed</span>{` Until ${timeRange[0]+' '+timeRange[1]}`}</div>}
+                {}
+            </div>
+        </div>
+        
+    )
+}
+
+export default RestaurantView;
```
Comment (MatthewVaccaro) [inline src/ui/views/restaurantView/restaurantView.jsx]: There is a lot of extra logic and extra variables being created that make this really hard to follow the logic of it all. The steps to break this into is
1. Find the current time and break it into helpful bitesize pieces
2. Convert today's and yesterdays restaurant hours into 24-hour military time ints
3. Compare times
    -- Compare if the current time is AM and yesterdays closing time is AM (Move on if restaurant was closed yesterday)
    -- Check if today's hours are closed
    -- Normal check: Current time is greater than open but less than closed
    -- Last check (For places that close on a late AM) if the current time is greater than open & closed hour is AM

The code I wrote on the backend
```suggestion
const dateConstructor = new Date();
const currentDate = dateConstructor.toString().split(" ");
const day = dateConstructor.getDay(); // Returns INT 0 - 6

const allDays = ["sun", "mon", "tue", "wed", "thu", "fri", "sat"];

const today = currentDate[0].toLowerCase();
const yesterday = day === 0 ? allDays[6] : allDays[day - 1]; // On Sunday, yestorday equals saturday

const time = currentDate[4].split(":");
const hour = parseInt(time[0] + time[1]); //[0] Select Hour | [1] Selects Minute

function isOpen(hoursObject) {
	const restaurantHoursToday = hoursObject[today].split(" "); // ["5:00", "AM", "11:00", "AM"]
	const restaurantHoursYesterday = hoursObject[yesterday].split(" "); // ["5:00", "AM", "11:00", "AM"]

	//NOTE: If current time & closed time is AM
	if (restaurantHoursYesterday[0] !== "Closed" && restaurantHoursYesterday[3] === "AM") {
		const yesterdayOpenHour = restaurantHoursYesterday[0].split(":");
		const yesterdayClosedHour = restaurantHoursYesterday[2].split(":");
		// Final check to know if the AM time is extending the day - if Open time is > closed time (11am = 1100 > 1am = 100)
		if (parseInt(yesterdayClosedHour[0] + yesterdayClosedHour[1]) < parseInt(yesterdayOpenHour[0] + yesterdayOpenHour[1])) {
			if (hour < parseInt(yesterdayClosedHour[0] + yesterdayClosedHour[1])) {
				return true;
			}
		}
	}
	//NOTE: Check if Closed
	if (restaurantHoursToday[0] === "Closed") return false;

	const restaurantOpenHours = restaurantHoursToday[0].split(":");
	const restaurantCloseHours = restaurantHoursToday[2].split(":");

	const openMeridian = restaurantHoursToday[1];
	const closedMeridian = restaurantHoursToday[3];

	const openHour =
		openMeridian === "AM"
			? parseInt(restaurantOpenHours[0] + restaurantOpenHours[1])
			: parseInt(restaurantOpenHours[0] + restaurantOpenHours[1]) + 1200;

	const closedHour =
		closedMeridian === "AM"
			? parseInt(restaurantCloseHours[0] + restaurantCloseHours[1])
			: parseInt(restaurantCloseHours[0] + restaurantCloseHours[1]) + 1200;

	//NOTE: Normal check: Current > open && Current < closed
	if (hour > openHour && hour < closedHour) {
		return true;
	}

	//NOTE: If current time is PM & closed time is AM
	if (hour > openHour && restaurantHoursToday[3] === "AM") {
		return true;
	}

	return false;
}
```

**Review by MatthewVaccaro — CHANGES_REQUESTED** [review]
We can sync on this on Sunday if you want. Let me know

File: src/ui/views/restaurantView/restaurantView.jsx (hunk 4 above)
Comment (mjedwards) [reply src/ui/views/restaurantView/restaurantView.jsx]: this function is checking the store's time. Depending on when the store opens or closes, it will convert the time to 24hr or leave it alone. I push the "AM/PM" to the array because I felt it would be easier to access it when I display the time in the UI/UX.  

Your suggested code would work. I will try it out. Thanks for the help!
